### PR TITLE
real time metrics specific to latest promoted versions

### PR DESCRIPTION
**Pull request by scrajesh**
1) Fetch metrics summary
2) Find recently promoted versions
3) Find real-time metrics specific to promoted versions
4) Aggregate both and send to the user

### CODE REVIEW

**Review by datvong-wm — CHANGES_REQUESTED** [review]
tests

File: electrode-ota-server-dao-mariadb/src/queries/PackageQueries.ts [hunk 1]
```diff
@@ -103,4 +103,6 @@ export const PackageQueries = {
                     WHERE id = ?`,
 
     updatePackageTime : `UPDATE package SET update_time = CURRENT_TIMESTAMP(3) WHERE id = ?`,
+
+    getMostRecentlyPromotedVersions : `SELECT app_version FROM package WHERE release_method IN ('Promote') ORDER BY update_time DESC LIMIT 3`,
```
Comment (datvong-wm) [inline electrode-ota-server-dao-mariadb/src/queries/PackageQueries.ts:107]: is this query going to be slow?

File: electrode-ota-server-dao-mariadb/src/queries/MetricQueries.ts [hunk 2]
```diff
@@ -27,4 +27,8 @@ export const MetricQueries = {
     getMetricsByStatusAfterSpecificTime : `SELECT count(*) as total, app_version, label, previous_deployment_key,
                     previous_label_or_app_version, status FROM metric WHERE deployment_id = ? AND create_time >= ?
                     GROUP BY status, label, app_version, previous_deployment_key, previous_label_or_app_version`,
+
+    getMetricsByStatusAfterSpecificTimeAndVersion : `SELECT count(*) as total, app_version, label, previous_deployment_key,
+                    previous_label_or_app_version, status FROM metric WHERE deployment_id = ? AND create_time >= ? AND app_version in (?,?,?)
+                    GROUP BY status, label, app_version, previous_deployment_key, previous_label_or_app_version`,
```
Comment (datvong-wm) [inline electrode-ota-server-dao-mariadb/src/queries/MetricQueries.ts:33]: is this query going to be slow?

File: electrode-ota-server-dao-mariadb/src/dao/MetricDAO.ts [hunk 3]
```diff
@@ -86,11 +88,23 @@ export default class MetricDAO extends BaseDAO {
 
     public static async metricsByStatusAfterSpecificTime(connection: IConnection, deploymentKey: string, specificDate: Date): Promise<MetricByStatusOutDTO[]> {
         const depResults = await MetricDAO.query(connection, DeploymentQueries.getDeploymentByKey, [deploymentKey]);
-
         if (!depResults || depResults.length === 0) {
             throw new Error("Not found. no deployment found for key [" + deploymentKey + "]");
         }
-        const metricResults = await MetricDAO.query(connection, MetricQueries.getMetricsByStatusAfterSpecificTime, [depResults[0].id, specificDate]);
+
+        const latestVersions = await PackageDAO.getMostRecentlyPromotedVersions(connection);
```
Comment (datvong-wm) [inline electrode-ota-server-dao-mariadb/src/dao/MetricDAO.ts:95]: what does all versions have to deal with the deploymentKey?

File: electrode-ota-server-dao-mariadb/src/queries/PackageQueries.ts (hunk 1 above)
Comment (scrajesh) [reply electrode-ota-server-dao-mariadb/src/queries/PackageQueries.ts:107]: Definitely not as the table has very lesser number(apporx 400) of data.

File: electrode-ota-server-dao-mariadb/src/dao/MetricDAO.ts (hunk 3 above)
Comment (scrajesh) [reply electrode-ota-server-dao-mariadb/src/dao/MetricDAO.ts:95]: Deployment Key doesn't play any role in metrics

File: electrode-ota-server-dao-mariadb/src/queries/MetricQueries.ts (hunk 2 above)
Comment (scrajesh) [reply electrode-ota-server-dao-mariadb/src/queries/MetricQueries.ts]: Initial run doesn't show any slowness. But on HTTP, yet to test on PROD.

File: electrode-ota-server-dao-mariadb/src/queries/PackageQueries.ts (hunk 1 above)
Comment (datvong-wm) [reply electrode-ota-server-dao-mariadb/src/queries/PackageQueries.ts:107]: You do know this is an ALL table scan.  Let's hope it never gets big.

File: electrode-ota-server-dao-mariadb/src/dao/MetricDAO.ts (hunk 3 above)
Comment (datvong-wm) [reply electrode-ota-server-dao-mariadb/src/dao/MetricDAO.ts:95]: Can this add this test:
App 1 versions:   1.0 2.0 3.0 4.0 5.0 6.0
App 2 versions: 1.0 2.0
All versions are PROMOTED.

What happens when I get metric data for App 2?
My understanding of this code is, it'll return versions [4.0, 5.0, 6.0].
Then your going to try to find metrics for App 2, versions [4.0, 5.0, 6.0]

If you understand this issue, please add a test for it.

File: electrode-ota-server-dao-mariadb/src/queries/MetricQueries.ts (hunk 2 above)
Comment (scrajesh) [reply electrode-ota-server-dao-mariadb/src/queries/MetricQueries.ts]: Running the query in PROD DB, following is the stats seen:

- With specific app-version the runtime is around `1.10s`

- Without involving app-version the runtime is around `1.40s`

File: electrode-ota-server-dao-mariadb/src/dao/MetricDAO.ts (hunk 3 above)
Comment (scrajesh) [reply electrode-ota-server-dao-mariadb/src/dao/MetricDAO.ts]: Tweaked this query to retrieve recently promoted versions of the app requested. Also added unit tests as above.

When metrics for App 1 is requested, it retrieves for versions [6.0, 5.0, 4.0] of App-1

When metrics for App 2 is requested, it retrieves for versions [2.0, 1.0] of App-2

File: electrode-ota-server-dao-mariadb/src/queries/MetricQueries.ts (hunk 2 above)
Comment (datvong-wm) [reply electrode-ota-server-dao-mariadb/src/queries/MetricQueries.ts]: timing is good.  using `explain` is better.

**Review by datvong-wm — APPROVED** [review]

**Pull request merged**